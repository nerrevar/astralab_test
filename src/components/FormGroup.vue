<template>
  <div class="form-group">
    <div class="form-group-label">
      <span class="form-group-label__title">
        {{ item.label }}
      </span>
      <span
        v-if="item.hint"
        class="form-group-label__hint"
        :title="item.hint"
      >
        ?
      </span>
    </div>
    <input
      :class="{
        'form-group__input': item.inputType !== 'password',
        'form-group__input_with-eye': item.inputType === 'password',
        'form-group__input_error': errorFields.indexOf(item.inputName) !== -1,
      }"
      :type="item.inputType"
      :name="item.inputName"
      ref="input"
      @input="validate($event.target.value)"
    />
    <span
      class="form-group__error"
      v-if="errorFields.indexOf(item.inputName) !== -1"
    >
      {{ item.errorMessage }}
    </span>
    <span
      class="form-group__error_hidden"
      v-else
    >
      &nbsp;
    </span>
  </div>
</template>

<script>
export default {
  name: 'FormGroup',
  props: {
    item: { required: true },
    errorFields: {
      type: Array,
      required: true,
    },
  },
  methods: {
    validate (str) {
      if (this.item.validationFunction)
        if (this.item.validationFunction(str))
          this.$emit('validate', this.item.inputName)
        else
          this.$emit('invalidate', this.item.inputName)
    },
    isValid () {
      if (this.item.validationFunction)
        if (this.item.validationFunction(this.$refs.input.value))
          return true
        else {
          this.$emit('invalidate', this.item.inputName)
          return false
        }
      return true
    },
    getData () {
      return { [this.item.inputName]: this.$refs.input.value }
    },
  },
}
</script>

<style lang="sass" scoped>
.form-group
  display: flex
  flex-flow: column nowrap
  padding: 12px 0

  &-label
    display: flex
    flex-flow: row nowrap
    justify-content: space-between
    padding: 0 20px

    &__title
      font-weight: 600
      font-size: 12px
      line-height: 20px

    &__hint
      +round
      width: 12px
      height: 12px
      text-align: center
      padding: 0
      color: #f4f4f4
      background-color: #cccbd2

  &__input
    +ellipse
    padding: 10px 2em

    &_error
      background-color: #ffe0e6

    &_with-eye
      +ellipse
      padding: 10px 2em

      &::after
        content: 'a'
        margin-left: -20px
        background-color: black
        width: 12px
        height: 12px

  &__error
    font-size: 10px
    line-height: 20px
    color: #ff7992
    padding: 0 20px

    &_hidden
      display: block
      width: 0
      height: 20px
</style>
