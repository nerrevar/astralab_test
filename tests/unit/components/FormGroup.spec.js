import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'

import FormGroup from '@/components/FormGroup.vue'

describe('FormGroup.vue input type - text', () => {
  const nameValidationFunction = str => str.trim() !== '' && str.replace(/[a-z ]+/i, '') === ''
  const vm = shallowMount(FormGroup, {
    propsData: {
      item: {
        label: 'Full name',
        inputType: 'text',
        inputName: 'name',
        errorMessage: 'Must contain only english letters',
        validationFunction: nameValidationFunction,
      },
      errorFields: [],
    },
  })

  it('renders', () => {
    expect(vm.element).toMatchSnapshot()
  })

  it('label text', () => {
    expect(vm.find('.form-group-label__title').text()).toBe('Full name')
  })

  it('label hint existing', () => {
    expect(vm.find('.form-group-label__hint').exists()).toBeFalsy()
  })

  it('error message visibility', () => {
    expect(vm.find('.form-group__error').exists()).toBeFalsy()
    vm.setProps({ errorFields: ['name'] })
    Vue.nextTick(() => {
      expect(vm.find('.form-group__error').exists()).toBeTruthy()
      expect(vm.find('.form-group__error').text()).toBe('Must contain only english letters')
    })
  })

  it('method validate on aaa', () => {
    vm.find('input').setValue('aaa')
    vm.find('input').trigger('input')
    expect(vm.emitted('validate')).toBeTruthy()
  })

  it('method validate on empty', () => {
    vm.find('input').setValue('')
    vm.find('input').trigger('input')
    expect(vm.emitted('invalidate')).toBeTruthy()
  })

  it('method isValid on aaa', () => {
    vm.find('input').setValue('aaa')
    expect(vm.vm.isValid()).toBeTruthy()
  })

  it('method isValid on empty', () => {
    vm.find('input').setValue('')
    expect(vm.vm.isValid()).toBeFalsy()
    expect(vm.emitted('invalidate')).toBeTruthy()
  })

  it('method isValid with empty validation function', () => {
    vm.vm.item.validationFunction = null
    expect(vm.vm.isValid()).toBeTruthy()
  })

  it('method getData on aaa', () => {
    vm.find('input').setValue('aaa')
    expect(vm.vm.getData()).toStrictEqual({ name: 'aaa' })
  })

  it('method getData on empty', () => {
    vm.find('input').setValue('')
    expect(vm.vm.getData()).toStrictEqual({ name: '' })
  })
})
