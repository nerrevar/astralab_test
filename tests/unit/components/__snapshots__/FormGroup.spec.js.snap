// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormGroup.vue input type - text renders 1`] = `
<div
  class="form-group"
>
  <div
    class="form-group-label"
  >
    <span
      class="form-group-label__title"
    >
      
      Full name
    
    </span>
     
    <!---->
  </div>
   
  <input
    class="form-group__input"
    name="name"
    type="text"
  />
   
  <span
    class="form-group__error_hidden"
  >
    
    Â 
  
  </span>
</div>
`;
